### ISSUE TITLE
Goals for "Syntax" Folder

### ISSUE THREAD

**Issue by JordanMartinez** [JordanMartinez/purescript-jordans-reference#36]
Purpose: 
- Explain and document basic to advance syntax using meta-language

Principles:
- Verify syntax by making it compile, so that code can be quickly updated for newer Purescript releases
- Make the folder referrable: keep all syntax-related code in the folder and exclude everything else
- Make the folder learnable: clearly and easily indicate pre-reqs for a page. In other words, how long can I read the pages in numerical order before I will need to stop and return to something later?

Outcome:
- New learners can gain enough knowledge/familiarity to read through and understand the code in the Hello World folder
- People can look at this folder to quickly remind themselves how to write some thing.

**Comment by JordanMartinez** [comment]
I recently put all the source files under one folder so that one could `pulp --psc-package build` / `pulp repl` it easily. However, the Module Syntax folder emits compiler warnings (because it is efficient) and we are choosing not to be efficient so that things are fully documented. These warnings add 'noise' to the compilation should one wish to experiment with the syntax a bit in the `01-Basic-Syntax` folder. As such, I think it should be moved into its own folder.

Moreover, since type-level programming allows one to write their own compiler warnings/errors, this should also go in its own folder to separate it from others.

Also, since I'm now documenting the type-level programming syntax, the module syntax does not currently document how to import/export kinds. As such, I think it should go after the type-level programming syntax folder. The FFI folder can stay in the same spot. Thus, I coould reorganize things to be:
```
11-Syntax/
  01-Warning-Free/
    01-Basic-Syntax
    02-Foreign-Function-Interface-Syntax
  02-Warnings-Emitted
    01-Type-Level-Programming-Syntax
    02-Module-Syntax
```
However, at that point, I might as well revert back to separating them out entirely as I don't know whether additional warnings may arise in the future (e.g. FFI).

**Comment by JordanMartinez** [comment]
As stated [here](https://functionalprogramming.slack.com/archives/C04NA444H/p1535067349000100) in reference to type-level programming:
> The constraint solver is just runtime for the type system

**Comment by JordanMartinez** [comment]
[This link](https://www.schoolofhaskell.com/school/to-infinity-and-beyond/pick-of-the-week/type-families-and-pokemon) explains type families using Pokemon as an analogy. 
TL;DR:
```haskell
class MyClass first second | first -> second where
  data DifferentForEachInstance (a :: *)
  function :: first -> first -> second
  otherFunction :: first -> DifferentForEachInstance a
```

**Comment by JordanMartinez** [comment]
[Document this syntax](https://github.com/justinwoo/quoted-record-property/blob/master/src/Main.purs)

**Comment by JordanMartinez** [comment]
[This link](https://cdepillabout.github.io/haskell-type-families-presentation/#/) explains the sameish idea explained in the below link much faster.

[This link](https://www.schoolofhaskell.com/school/to-infinity-and-beyond/pick-of-the-week/type-families-and-pokemon) explains type families using Pokemon as an analogy.

**Comment by JordanMartinez** [comment]
Type-Level syntax has now been documented. There's still a few things to do:
- [x] ~Give examples that can be run in a REPL to demonstrate how TLP is possible.~ Decided to move this to the Hello World folder by using custom type errors
- [x] Demonstrate how to write multi-arg functions
- [x] Figure out how to write the functional dependencies correctly and check whether I've written them correctly. (see #92)
- [x] Update the Module syntax to now import/export kinds and their instances

I also think the explanations for Prim should be moved to the Hello World folder. I can explain how to write custom type errors in that folder since custom compiler warnings don't appear in the REPL as paf31 explained.

**Comment by JordanMartinez** [comment]
The FFI syntax could be updated to show more examples, but it's enough for me for now. I'll probably go through it some more when I encounter a need for that.
- Add more examples of FFI syntax...?

**Comment by JordanMartinez** [comment]
I no longer think I should add more examples of FFI syntax. Rather, the heart of the issue is more like "What are best practices for implementing a binding for some JS library?" That itself is something that should fall under Design Patterns, not Syntax.

Having said that, this issue, AFAIK, is finished.
